File: app/src/androidTest/java/com/bbtest/utils/CommonUtil.java
package com.bbtest.utils;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Random;

public class CommonUtil {

    public static int getFirstNum(String str) {
        int num = -1;
        String tmpNum = "";
        boolean isStart = false;
        for (int i = 0; i < str.length(); i++) {
            if (Character.isDigit(str.charAt(i))) {
                isStart = true;
                tmpNum += str.charAt(i);
            } else {
                if (isStart) {
                    break;
                }
            }
        }
        if (!tmpNum.equals("")) {
            num = Integer.parseInt(tmpNum);
        }
        return num;
    }

    public static boolean isNumer(String str) {
        for (int i = 0; i < str.length(); i++) {
            if (!Character.isDigit(str.charAt(i))) {
                return false;
            }
        }
        return true;
    }

    /**
     * 方案1：Double.valueOf(String.format("%.1f", 0.5624))
     * 方案2：DecimalFormat df = new DecimalFormat("0.00");
     * Double.valueOf(df.format(0.85621));
     * 方案3：BigDecimal bd = new BigDecimal(0.24684136);
     * bd.setScale(2, BigDecimal.ROUND_HALF_UP).doubleValue();
     */
    public static double keepDecimalPoint(double number, int point) {
        return Double.valueOf(String.format("%." + point + "f", number));
    }

    /**
     * 随机生成字符串
     */
    public static String randomStr(int length) {
        String alphabetsInUpperCase = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        String alphabetsInLowerCase = "abcdefghijklmnopqrstuvwxyz";
        String numbers = "0123456789";

        String allCharacters = alphabetsInLowerCase + alphabetsInUpperCase + numbers;
        StringBuffer randomString = new StringBuffer();
        for (int i = 0; i < length; i++) {
            int randomIndex = new Random().nextInt(allCharacters.length());
            randomString.append(allCharacters.charAt(randomIndex));
        }
        return randomString.toString();
    }

    /**
     * 生成 [m,n] 的数字
     */
    public static int randomInt(int startNum, int endNum) {
        return new Random().nextInt(endNum - startNum + 1) + startNum;
    }

    public static String getCurYear() {
        return new SimpleDateFormat("yyyy").format(new Date());
    }

    public static String getCurTime() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMddHHmmss");
        return dateFormat.format(new Date());
    }

    public static String getCurTimeForFile() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyyMMdd_HHmmss");
        return dateFormat.format(new Date());
    }

    public static String getCurTimeForLog() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss:SSS");
        return dateFormat.format(new Date());
    }

    public static String getExceptionMsg(Exception e) {
        String exceptionMsg = "";
        try {
            String msg = e.getMessage();
            if (msg != null) {
                exceptionMsg += "   " + msg + "\n";
            }
            StackTraceElement[] stacks = e.getStackTrace();
            if (stacks != null) {
                for (int i = 0; i < stacks.length; i++) {
                    exceptionMsg += "   " + stacks[i].toString() + "\n";
                }
            }
            Throwable cause = e.getCause();
            if (cause != null) {
                exceptionMsg += "   Caused by: " + cause.getMessage() + "\n";
            }
        } catch (Exception e1) {
            e1.printStackTrace();
        }
        return exceptionMsg;
    }

    public static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/androidTest/kotlin/com/bbtest/utils/CommonUtil.kt
package com.bbtest.utils

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Random

object CommonUtil {
    private const val FILE_TIME_PATTERN = "yyyyMMdd_HHmmss"
    private const val LOG_TIME_PATTERN = "yyyy-MM-dd HH:mm:ss:SSS"
    private const val FULL_TIME_PATTERN = "yyyyMMddHHmmss"
    private const val YEAR_PATTERN = "yyyy"

    @JvmStatic
    fun getFirstNum(str: String): Int {
        val tmpNum = StringBuilder()
        var isStart = false
        for (char in str) {
            if (char.isDigit()) {
                isStart = true
                tmpNum.append(char)
            } else if (isStart) {
                break
            }
        }
        return tmpNum.toString().toIntOrNull() ?: -1
    }

    @JvmStatic
    fun isNumer(str: String): Boolean {
        for (char in str) {
            if (!char.isDigit()) {
                return false
            }
        }
        return true
    }

    @JvmStatic
    fun keepDecimalPoint(number: Double, point: Int): Double {
        return String.format("%.${point}f", number).toDouble()
    }

    @JvmStatic
    fun randomStr(length: Int): String {
        val allCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        val randomString = StringBuilder()
        val random = Random()
        for (i in 0 until length) {
            val randomIndex = random.nextInt(allCharacters.length)
            randomString.append(allCharacters[randomIndex])
        }
        return randomString.toString()
    }

    @JvmStatic
    fun randomInt(startNum: Int, endNum: Int): Int {
        return Random().nextInt(endNum - startNum + 1) + startNum
    }

    @JvmStatic
    fun getCurYear(): String {
        return SimpleDateFormat(YEAR_PATTERN).format(Date())
    }

    @JvmStatic
    fun getCurTime(): String {
        return SimpleDateFormat(FULL_TIME_PATTERN).format(Date())
    }

    @JvmStatic
    fun getCurTimeForFile(): String {
        return SimpleDateFormat(FILE_TIME_PATTERN).format(Date())
    }

    @JvmStatic
    fun getCurTimeForLog(): String {
        return SimpleDateFormat(LOG_TIME_PATTERN).format(Date())
    }

    @JvmStatic
    fun getExceptionMsg(e: Exception): String {
        val exceptionMsg = StringBuilder()
        try {
            e.message?.let { msg ->
                exceptionMsg.append("   ").append(msg).append('\n')
            }
            e.stackTrace.forEach { stack ->
                exceptionMsg.append("   ").append(stack).append('\n')
            }
            e.cause?.let { cause ->
                exceptionMsg.append("   Caused by: ").append(cause.message).append('\n')
            }
        } catch (e1: Exception) {
            e1.printStackTrace()
        }
        return exceptionMsg.toString()
    }

    @JvmStatic
    fun sleep(millis: Long) {
        try {
            Thread.sleep(millis)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
